From 0ff6705cde96601543999d1eaf8c0568450b5537 Mon Sep 17 00:00:00 2001
From: Dan Nica <dan.nica@nxp.com>
Date: Fri, 2 Oct 2020 20:55:36 +0300
Subject: [PATCH] sja1110_init: fix various issues

This commit also resolves some issues:
	* The RCNT field of the SPI message was always set to the
	  reserved value of 0.
	* The address of the RESET_CTRL_0 register was defined
	  incorrectly.
	* The dev_id_idx variable was being used un-initialized
	  in sja1110_pre_switch_upload(...)

Signed-off-by: Ryder Gong <ryder.gong@nxp.com>
Signed-off-by: Dan Nica <dan.nica@nxp.com>
Signed-off-by: Radu Pirea (NXP OSS) <radu-nicolae.pirea@oss.nxp.com>
---
 sja1110_init.c | 17 ++++++++++-------
 sja1110_init.h |  5 +++--
 2 files changed, 13 insertions(+), 9 deletions(-)

diff --git a/sja1110_init.c b/sja1110_init.c
index b8d4eaf..fcf4818 100644
--- a/sja1110_init.c
+++ b/sja1110_init.c
@@ -45,7 +45,8 @@
 /*******************************************************************************
  * Global variables
  ******************************************************************************/
-static const u32 device_id_list[] = {SJA1110_VAL_DEVICEID};
+static const u32 device_id_list[] = {SJA1110_VAL_DEVICEID,
+				     SJA1110_VAL_DEVICEID | 1};
 #define NUM_DEVICE_IDS ARRAY_SIZE(device_id_list)
 
 struct sja1110_state g_state;
@@ -91,7 +92,7 @@ static u32 sja1110_read_reg(struct sja1110_priv *sja1110, u32 reg_addr)
 
 	cmd[0] = cpu_to_le32(CMD_ENCODE_RWOP(CMD_RD_OP) |
 			     CMD_ENCODE_ADDR(reg_addr) |
-			     CMD_ENCODE_WRD_CNT(0));
+			     CMD_ENCODE_WRD_CNT(1));
 	cmd[1] = 0;
 
 	ret = sja1110_spi_transfer(sja1110, (u8*)cmd, (u8*)resp, 8, 8);
@@ -205,13 +206,14 @@ static int sja1110_pre_switch_upload(struct sja1110_priv *sja1110,
 				     const u8 *cfg_data, int cfg_size)
 {
 	int dev_id_idx, ret = -EINVAL;
-	u32 registerValue, cfg_device_id;
+	u32 regval, cfg_device_id;
 
 	/* read the device id */
-	registerValue = sja1110_read_reg(sja1110, D_ID_ADDR);
-	if (sja1110_check_device_id(registerValue) < 0) {
+	regval = sja1110_read_reg(sja1110, D_ID_ADDR);
+	dev_id_idx = sja1110_check_device_id(regval);
+	if (dev_id_idx < 0) {
 		dev_err(&sja1110->spi->dev,
-			"Device id 0x%x is not supported!\n", registerValue);
+			"Device id 0x%x is not supported!\n", regval);
 		goto out;
 	}
 
@@ -443,7 +445,8 @@ static int sja1110_uc_parse_status_many(struct sja1110_priv *sja1110,
 			goto out;
 		}
 
-		if (uc_status.status != DOWNLOADING ||
+		if ((uc_status.status != DOWNLOADING &&
+		     uc_status.status != VALIDATING_IMG_META_DATA) ||
 		    uc_status.err_code != NO_ERROR) {
 			dev_err(&sja1110->spi->dev,
 				"[%s] Upload error detected in status message %d (status=0x%x,err=0x%x)\n",
diff --git a/sja1110_init.h b/sja1110_init.h
index d8f38ec..88d0f1f 100644
--- a/sja1110_init.h
+++ b/sja1110_init.h
@@ -56,7 +56,8 @@ enum uc_status {
 	WAITING      = 0x33,
 	DOWNLOADING  = 0x34,
 	VERIFYING    = 0x35,
-	COMPLETED    = 0x36
+	COMPLETED    = 0x36,
+	VALIDATING_IMG_META_DATA = 0x37,
 };
 enum uc_err_code {
 	SFDP_FAILURE                = 0x51,
@@ -91,7 +92,7 @@ enum uc_err_code {
 #define CONFIG_START_ADDRESS   (0x20000UL)  /**< Start Address of the configuration */
 #define D_ID_ADDR              (0x0UL)      /**< Address of the deviceId register */
 #define CONFIGURATIONFLAG_ADDR (0x1UL)      /**< Address of the configurationFlags register */
-#define R_CTRL_ADDR            (0x1C6000UL) /**< Address of the resetCtrl register */
+#define R_CTRL_ADDR            (0x1C6100UL) /* RESET_CTRL_0 */
 #define RESET_CTRL_COLDRESET   BIT(5)
 #define RESET_DELAY_US         50
 #define CGU_SPI_BASE_ADDR      (0x1C6400UL)
-- 
2.34.1

